public/hooks/use_chat_actions.tsx

```diff
@@ -3,83 +3,116 @@
-        body: JSON.stringify({
-          conversationId: chatContext.conversationId,
-          ...(!chatContext.conversationId && { messages: chatState.messages }), // include all previous messages for new chats
-          input,
-        }),
-        query: core.services.dataSource.getDataSourceQuery(),
-      });
-      if (abortController.signal.aborted) return;
-      // Refresh history list after new conversation created if new conversation saved and history list page visible
-      if (
-        !chatContext.conversationId &&
-        response.conversationId &&
-        core.services.conversations.options?.page === 1 &&
-        chatContext.selectedTabId === TAB_ID.HISTORY
-      ) {
-        core.services.conversations.reload();
-      }
-      chatContext.setConversationId(response.conversationId);
-      // set title for first time
-      if (response.title && !chatContext.title) {
-        chatContext.setTitle(response.title);
+  const send = (input: IMessage): Promise<void> =>
+    new Promise(async (resolve) => {
+      const abortController = new AbortController();
+      abortControllerRef = abortController;
+      chatStateDispatch({ type: 'send', payload: input });
+      let chunk$ = new BehaviorSubject<StreamChunk | undefined>(undefined);
+      try {
+        const fetchResponse = await core.services.http.post(ASSISTANT_API.SEND_MESSAGE, {
+          // do not send abort signal to http client to allow LLM call run in background
+          body: JSON.stringify({
+            conversationId: chatContext.conversationId,
+            ...(!chatContext.conversationId && { messages: chatState.messages }), // include all previous messages for new chats
+            input,
+          }),
+          query: core.services.dataSource.getDataSourceQuery(),
+          asResponse: true,
+        });
+        chunk$ = await getConsumedChunk$FromHttpResponse({
+          fetchResponse,
+          abortController,
+        });
+      } catch (error) {
+        resolve(undefined);
+        if (abortController.signal.aborted) {
+          return;
+        }
+        chatStateDispatch({ type: 'error', payload: error });
       }
-      /**
-       * Remove messages that do not have messageId
-       * because they are used for displaying loading state
-       */
-      chatStateDispatch({
-        type: 'receive',
-        payload: {
-          messages: chatState.messages.filter((item) => item.messageId),
-          interactions: chatState.interactions,
-        },
-      });
 
-      /**
-       * Patch messages and interactions based on backend response
-       */
-      chatStateDispatch({
-        type: 'patch',
-        payload: {
-          messages: response.messages,
-          interactions: response.interactions,
-        },
-      });
-    } catch (error) {
-      if (abortController.signal.aborted) return;
-      chatStateDispatch({ type: 'error', payload: error });
-    }
-  };
+      chunk$.subscribe(
```

Comment:
Combine the text response and the streaming response in the same observable is a little bit complicated to me. How about we check the response type outside and use the JSON response directly?

Reply:
Make sense, the hook is a little bit complicated now.